src/main/java/uk/gov/hmcts/reform/slc/services/steps/maptoletter/LetterMapper.java

```diff
@@ -15,17 +21,33 @@
 public class LetterMapper {
 
     private static final ObjectMapper objectMapper = new ObjectMapper();
+    private static final Validator validator = Validation.buildDefaultValidatorFactory().getValidator();
+
+    @Autowired
+    private AppInsights insights;
+
 
     public Letter from(IMessage msg) {
+        byte[] body = msg.getBody();
+        String messageId = msg.getMessageId();
+
         try {
-            Letter letter = objectMapper.readValue(msg.getBody(), Letter.class);
-            if (letter != null) {
+            Letter letter = objectMapper.readValue(body, Letter.class);
+            Set<ConstraintViolation<Letter>> violations = validator.validate(letter);
+
+            if (violations.isEmpty()) {
+                insights.trackMessageMappedToLetter(messageId, letter.service, letter.template, body.length);
+
                 return letter;
             } else {
-                throw new InvalidMessageException("Empty message " + msg.getMessageId());
-            }
+                insights.trackMessageMappedToNull(messageId, body.length);
 
+                // can work on message build from violations
+                throw new InvalidMessageException("Invalid message body");
+            }
         } catch (IOException exc) {
+            insights.trackMessageNotMapped(messageId, body.length);
```

Comment:
Is this an actual 'event'? should it not be tracked as error? telemetry client has `.trackException()` method.

Reply:
Not tracking any exception raised by us. Only from third party